File: gfg/easy/find-missing-and-repeating2512/solution.js
/**
 * @param {number[]} arr
 * @returns {number[]}
 */
class Solution {
    findTwoElement(arr) {
        const n = arr.length;

        let xorVal = 0;

        // XOR all array elements
        for (let num of arr) {
            xorVal ^= num;
        }

        // XOR numbers from 1 to n
        for (let i = 1; i <= n; i++) {
            xorVal ^= i;
        }

        // Get rightmost set bit
        let setBit = xorVal & -xorVal;

        let x = 0;
        let y = 0;

        // Divide array elements into two groups
        for (let num of arr) {
            if (num & setBit) {
                x ^= num;
            } else {
                y ^= num;
            }
        }

        // Divide numbers 1 to n into two groups
        for (let i = 1; i <= n; i++) {
            if (i & setBit) {
                x ^= i;
            } else {
                y ^= i;
            }
        }

        // Find which is repeating and which is missing
        let count = 0;
        for (let num of arr) {
            if (num === x) count++;
        }

        if (count === 2) {
            return [x, y];
        } else {
            return [y, x];
        }
    }
}

File: gfg/easy/find-missing-and-repeating2512/solution.py
class Solution:
    def findTwoElement(self, arr):
        n = len(arr)

        xor_val = 0

        # XOR all array elements and numbers from 1 to n
        for num in arr:
            xor_val ^= num

        for i in range(1, n + 1):
            xor_val ^= i

        # Find rightmost set bit
        set_bit = xor_val & -xor_val

        x = 0
        y = 0

        # Divide numbers into two groups
        for num in arr:
            if num & set_bit:
                x ^= num
            else:
                y ^= num

        for i in range(1, n + 1):
            if i & set_bit:
                x ^= i
            else:
                y ^= i

        # Determine which one is repeating
        if x in arr:
            return [x, y]
        else:
            return [y, x]